### DISCUSSION TITLE
(GTFS vs GTFS-RT) are stop times endpoints updated in real-time?

### DISCUSSION THREAD

**Discussion by jakswa** [transitland/transitland#348] [Q&A]
I was exploring using this API as a replacement for GTFS + GTFS-RT ingestion. But it is not clear whether transit.land tries to keep `stop_times` updated based on trip updates that can appear in GTFS-RT endpoints. For example a bus might be stuck in traffic and be behind. In this case, I _think_ the design is that a trip update will be issued in the GTFS-RT for that agency. Does transit.land serve this use case?

**Comment by drewda** [comment]
The public Transitland APIs currently only show schedule data from static GTFS feeds. But you are right that the v2 APIs are designed to also have places for GTFS Realtime data. We have been working on this and will have more to share in the future about static + real-time data available in the public Transitland v2 APIs.

**Reply by jakswa** [reply]
thank you, that makes sense

**Comment by patricksheehan** [comment]
Hi there, I wanted to check on this. I'm interested in the same use-case and just checked to see that the GTFS-RT feed I'm interested (BART) was last fetched 4 days ago. I'm assuming to have up-to-date RT delays you would have to fetch every minute. Is that happening behind the scenes or is GTFS-RT ingestion still in progress?

**Reply by irees** [reply]
hi @patricksheehan 

Yes, since this discussion was created, we've started including GTFS-RT information in both the REST and GraphQL API.

For example, a request to the [stop departures api](https://www.transit.land/documentation/rest-api/departures):
https://transit.land/api/v2/rest/stops/f-sf~bay~area~rg:FTVL/departures?include_alerts=true&apikey=YOURKEY

```     
    -- snip --
    "departures": [
        {
          "arrival": {
            "delay": 199,
            "estimated": "19:29:42",
            "estimated_utc": "2022-12-22T03:29:42Z",
            "scheduled": "19:20:00",
            "uncertainty": 30
          },
          "departure": {
            "delay": 199,
            "estimated": "19:30:19",
            "estimated_utc": "2022-12-22T03:30:19Z",
            "scheduled": "19:20:00",
            "uncertainty": 30
          },
          -- snip --
```

The "estimated" and "estimated_utc" values are based on GTFS-RT trip updates for this feed; "scheduled" comes from the static GTFS for this trip. Additionally, when `include_alerts=true` is included in the request, service alerts will be added to stops/routes/agencies/trips in the response where applicable.

**Reply by hannashimona** [reply]
@irees, how exciting! How will we know when GTFS-RT is available for a given feed? 

The documentation seems to suggest that MTA subway lines now have RT information available, but `https://transit.land/api/v2/rest/stops/361720555/departures?include_alerts=true&apikey=MYKEY `doesn't include RT info, and `https://transit.land/api/v2/rest/routes/f-mta~nyc~rt~subway~g?apikey=MYKEY` doesn't yield any info at all, while `https://transit.land/api/v2/rest/routes/r-dr5r-g?apikey=MYKEY` seems to be the static feed.

**Reply by kuanb** [reply]
Is there further documentation about the API response? I seem to recall there being some, though I cannot find it now. I see a [blog post](https://www.transit.land/news/2019/09/26/real-time-additions) which right now is missing the source markdown file (`news/2019-09-26-real-time-additions.md`) that would (possibly?) describe the structure of the response, though it may be too old.

I am specifically trying to determine the frequency of ghost buses (scheduled but unserved) for routes. I am picking arbitrary stops along the midpoint of the route in each direction and then seeking seeing what schedule adherence is like.

When I see cases like the following for an arrival:
```
{'delay': None,
 'estimated': None,
 'estimated_utc': None,
 'scheduled': '09:54:30',
 'uncertainty': None}
```
...is it reasonable to assume that a `None` value indicates that no live vehicle location was reported within a "reasonable" window around that scheduled time?

I ask because I looked at a single Onestop ID, such as `s-9q9p4jjhyw-oaklandav~montevistaav`, with date `2023-06-21` and looked specifically at the keys `scheduled` and `arrived` to see if both were present. I separated by whether they were/were not and noted that all but 4 were missing RT arrival data (I know there were some missing buses that day, but I don't think that many):

```
{'missed': [{'scheduled': '05:59:30', 'estimated': None},
  {'scheduled': '06:14:30', 'estimated': None},
  {'scheduled': '06:29:30', 'estimated': None},
  ...
  {'scheduled': '22:43:48', 'estimated': None},
  {'scheduled': '23:03:48', 'estimated': None}],
 'served': [{'scheduled': '09:14:30', 'estimated': '09:16:16'},
  {'scheduled': '09:17:24', 'estimated': '09:17:13'},
  {'scheduled': '09:34:30', 'estimated': '10:01:26'},
  {'scheduled': '09:54:30', 'estimated': '09:56:16'}]}
```

**Reply by dkter** [reply]
I'd also like to know more about this: for example, on some requests I see the `estimated` departure time is in the past (while the scheduled is in the future). Should I interpret this to mean the vehicle has already departed? How certain is this (i.e. is this because the GTFS-RT feed has reported that the vehicle has passed this stop, or just estimated based on other data)?

**Comment by babangidazachariah** [comment]
I want to repeatedly get trip updates and vehicle position from GTFS-RT. How can I get it done in python using transit.land API

**Comment by rmattich** [comment]
I have been testing gtfs-rt on a stop and for the most part is working well. However, I found that each night about 11pm local it returns no data and then  start getting results again sometime in the morning. I compare it to other applications (one bus away, connect point sign for the stop during the times I get nothing back and they both report correct. Is there something that could explain this? Below is a sample URL I'm testing.

https://transit.land/api/v2/rest/stops/s-c23jchdvek-internationalblvd~s180thst/departures?limit=3&apikey=<MY_API>'
